File: backend/responses/queueResponse.py
from typing import List
from pydantic import BaseModel

from backend.db.associationTables.user_queue_songs import QUEUE_LIST_TYPE_TYPE
from backend.responses.general.songWithAlbum import RockItSongWithAlbumResponse


class QueueResponseItemList(BaseModel):
    type: QUEUE_LIST_TYPE_TYPE
    publicId: str


class QueueResponseItem(BaseModel):
    song: RockItSongWithAlbumResponse
    queueSongId: int
    list: QueueResponseItemList


class QueueResponse(BaseModel):
    currentQueueSongId: int | None
    queue: List[QueueResponseItem]

File: frontend/responses/queueResponse.ts
import * as z from "zod";
import { RockItSongWithAlbumResponse } from "./rockItSongWithAlbumResponse";

import { QueueListTypes } from "@/types/rockIt";

export const QueueResponse = z.object({
    currentQueueSongId: z.number().nullable(),
    queue: z.array(
        z.object({
            song: RockItSongWithAlbumResponse,
            queueSongId: z.number(),
            list: z.object({
                type: z.enum(QueueListTypes),
                publicId: z.string(),
            }),
        })
    ),
});

export type QueueResponse = z.infer<typeof QueueResponse>;
